### PR TITLE
Add support for css, keyframes and injectGlobal keywords

### PR DESCRIPTION
**Pull request by zanza00**
This is my fix for #2330 

I also added the keyword `keyframes` that is used in the [same way](https://www.styled-components.com/docs/api#helpers) as `css`

The AST for them is slightly different since they are not `MemberExpression` but simple `Identifier`

### CODE REVIEW

File: src/multiparser.js [hunk 1]
```diff
@@ -294,16 +294,22 @@ function isStyledJsx(path) {
  * styled.button`color: red`
  * or
  * Foo.extend`color: red`
+ * or
```
Comment (vjeux) [inline src/multiparser.js]: could you remove all the `or`, they are not very useful now that we have 4 different ones :p

**Comment by vjeux** [comment]
lgtm, I'll merge it once you tweak the comments and travis is green :)

File: src/multiparser.js (hunk 1 above)
Comment (zanza00) [inline src/multiparser.js]: Yeah that's easy :)

**Comment by zanza00** [comment]
I noticed that the test fails because it's trying to compare the formatted output with the unformatted input. I have the same problem on my machine but I don't understand why it's happening :disappointed:

**Comment by vjeux** [comment]
Probably want to update clean-ast.js in order to not ensure the ast is the same: https://github.com/prettier/prettier/blob/master/src/clean-ast.js#L133-L151 Sorry I should have thought about that!

**Comment by zanza00** [comment]
Fair enough, I will update the PR then

**Comment by zanza00** [comment]
I noticed another keyword that styled components uses [injectglobal](https://www.styled-components.com/docs/api#injectglobal)

If it's okay I will add it to this PR

**Comment by vjeux** [comment]
It's fine, do it :)

**Comment by zanza00** [comment]
Done :smiley:

**Comment by vjeux** [comment]
Thanks!

**Comment by bakkot** [comment]
Bit late for it, but I object to this very strongly - while it's reasonable to assume that something like `styled.button ` is actually styled-components and therefore safe to reformat, this is not at all a guarantee for something like `keyframes` or `css`. I don't want Prettier messing with my AST just because I had a template tag named `keyframes` :(

More broadly, Prettier should support *JavaScript*, not "JavaScript as written by some subset of the community using these specific tools". It seems fine to add features to support those other tools, but not to build in the assumption that those other tools are the only things which exist, and e.g. no code formatted by Prettier will ever have a tag named `keyframes` which isn't from... whatever package this PR assumes such a tag is from.

**Comment by vjeux** [comment]
Actually, I agree with @bakkot. `css` is fine but `keyframe` and `injectGlobals` are very style-component specific. We should probably not support them inside of prettier-core. Do you mind sending a PR to remove them?

For your concerns @bakkot, if the expression doesn't parse, then we're going to print it as is. So hopefully it should only change the whitespaces and not introduce any breakage if you used CSS and not touch it if it's not CSS.

**Comment by bakkot** [comment]
@vjeux Can't for a while yet, but I'll try to get to it when I get home, if it's still up.

Side note: I'd love for these mappings of tag -> formatter to be configurable rather than hardcoded, as much as I normally prefer to [resist adding configuration](https://github.com/prettier/prettier/issues/40). This came up a long time ago, back when formatting template literals was first getting discussed, and now that Prettier has support for multiple parsers it's probably time to raise it again.

**Comment by vjeux** [comment]
I meant @zanza00 for the PR :)

For the configurability of those, I agree we'll need to find something more versatile. Right now we hardcode them for popular projects but we should figure out a better way.

**Comment by zanza00** [comment]
@vjeux to be clear
 -keep `css` 
 -remove `injectGlobals` and `keyframes`

and for configuration I think that one good option is to tell prettier what libraries are you using.

In this example I would configure prettier with styled-components so it will transform `injectGlobals` and `keyframes`

**Comment by vjeux** [comment]
@zanza00: correct.

Let's think about the best way to configure things in a separate pr :)

**Comment by janosh** [comment]
> Let's think about the best way to configure things in a separate pr :)

@vjeux Did that ever happen? `styled-components` v4 introduced `createGlobalStyle` which is not formatted by `prettier`. Any way I can configure that?

**Comment by vjeux** [comment]
@janosh this hasn't. If someone (you?) can send a similar PR as #2337 for `createGlobalStyle` then it'll work.

**Comment by janosh** [comment]
@vjeux Will it be accepted, though? Based on [your comment from 28 Jun 2017](https://github.com/prettier/prettier/pull/2337#issuecomment-311803127), I was under the impression that it was a conscious decision not to support `createGlobalStyle` by default and instead leave that to user configuration.
> Actually, I agree with @bakkot. css is fine but keyframe and injectGlobals are very style-component specific. We should probably not support them inside of prettier-core.

If you think the rise in `styled-component`'s popularity since mid 2017 now makes it worthwhile to support it by default, I'm happy to submit a PR.

**Comment by vjeux** [comment]
@janosh I'll accept it (make sure you @vjeux somewhere on the PR). Since we already added all the other ones in prettier, it doesn't make sense to leave that last one out.

It's not ideal and we probably want to do something about it at some point, but it'll be a longer discussion and in the meantime we should properly support the things we set out to, styled-components in this case.

**Comment by janosh** [comment]
Are you opposed to refactoring this
```js
ast.tag.name === "gql" ||
          ast.tag.name === "graphql" ||
          ast.tag.name === "css" ||
          ast.tag.name === "md" ||
          ast.tag.name === "markdown" ||
          ast.tag.name === "html"
```
to this?
```js
["gql", "graphql", "css", "md",  "markdown", "html"].includes(ast.tag.name)
```

**Comment by 19majkel94** [comment]
@vjeux Maybe prettier should support plugins that enhance the formatting with some specific rules, like `keyframes` or `createGlobalStyle` for `styled-components`?

**Comment by vjeux** [comment]
We may want to have that, but building a good plugin system is one of the hardest thing to do, so a lot of thoughts need to be put into it.

**Comment by xeoneux** [comment]
Can a comment next to the template literal help in this case?
Suppose, if I mark the `injectGlobal` line with `// prettier-css` or something similar, it formats the following string as CSS?

**Comment by janosh** [comment]
@vjeux Could you point me at where I need to make changes?

I added a test for `createGlobalStyle` in `tests/template-literals/styled-components-with-expressions.js` and included it in a check in `src/language-js/clean.js` but I'm not sure, where `prettier` actually identifies `TaggedTemplateExpression`s. I assume that's where I should add some logic to identify `createGlobalStyle`?

**Pull request merged**